sdk/src/main/scala/io/horizen/account/state/StateTransition.scala

```diff
@@ -23,63 +38,45 @@ class StateTransition(
    */
   @throws(classOf[InvalidMessageException])
   @throws(classOf[ExecutionFailedException])
-  def transition(msg: Message): Array[Byte] = {
+  def transition(): Array[Byte] = {
     // do preliminary checks
     preCheck(msg)
     // save the remaining block gas before any changes
     val initialBlockGas = blockGasPool.getGas
     // create a snapshot before any changes are made
-    val initialRevision = view.snapshot
+    val rollback = view.snapshot
+    var skipRefund = false
+    // allocate gas for processing this message
+    val gasPool = buyGas(msg)
```

Comment:
It is safe because the exceptions thrown by `buyGas` are before any changes to the blockGasPool or the state. `tracer.foreach(_.CaptureTxStart(gasPool.initialGas))` shouldn't throw any exception. However, if things changed in the future, maybe it would be necessary to put `buyGas` under the `try/catch`